rust/migration/src/m20220101_000001_create_blob_tables.rs

```diff
@@ -0,0 +1,128 @@
+use sea_orm_migration::prelude::*;
+
+#[derive(DeriveMigrationName)]
+pub struct Migration;
+
+#[async_trait::async_trait]
+impl MigrationTrait for Migration {
+    async fn up(&self, manager: &SchemaManager) -> Result<(), DbErr> {
+        manager
+            .create_table(
+                Table::create()
+                    .table(BlobStore::Table)
+                    .col(
+                        ColumnDef::new(BlobStore::Id)
+                            .integer()
+                            .not_null()
+                            .auto_increment()
+                            .primary_key(),
+                    )
+                    .col(ColumnDef::new(BlobStore::Type).string().not_null())
+                    .to_owned(),
+            )
+            .await?;
+
+        // Insert basic backing store since it must always exist.
+        let insert = Query::insert()
+            .into_table(BlobStore::Table)
+            .columns([BlobStore::Id, BlobStore::Type])
+            .values_panic([1.into(), "LocalBlobStore".into()])
+            .to_owned();
+
+        manager.exec_stmt(insert).await?;
+
+        manager
+            .create_table(
+                Table::create()
+                    .table(Blob::Table)
+                    .col(
+                        ColumnDef::new(Blob::Id)
+                            .integer()
+                            .not_null()
+                            .auto_increment()
+                            .primary_key(),
+                    )
+                    .col(ColumnDef::new(Blob::Key).string().not_null())
+                    .col(ColumnDef::new(Blob::StoreId).integer().not_null())
+                    .foreign_key(
+                        ForeignKeyCreateStatement::new()
+                            .name("fk-store_id-blob_store")
+                            .from_tbl(Blob::Table)
+                            .from_col(Blob::StoreId)
+                            .to_tbl(BlobStore::Table)
+                            .to_col(BlobStore::Id)
+                            // All blobs for a given store must be manually
+                            // deleted before the blob store may be deleted.
+                            .on_delete(ForeignKeyAction::Restrict),
```

Comment:
What are your thoughts on restrict vs cascade? What about specifically for blob store?

Reply:
In the comments I gave my core justification. But in general cascade is pretty dangerous.

In this case, if you deleted a BlobStore, it would cascade to the Blobs, which would cascade to the OutputFiles, which would cascade to the Jobs.

To me that really isn't acceptable especially since we have a clear workflow for deleting all those things 

1) Job eviction deletes a Job which cascades to its OutputFiles 
2) Blob eviction later sees those blobs with no references so it triggers proper deletion using the backing blob store API then deletes the entry in the blob table
3) We can then manually remove the decommissioned BlobStore

If we make it cascade we will be guaranteeing that blobs get orphaned in the remote store without really gaining anythign in the API